app/controllers/submission_viewing_events_controller.rb

```diff
@@ -147,14 +139,150 @@ def getTimingDetails
+      new = LocalSubmittedContent.new map_id: map_id,
+                                      round: round,
+                                      link: link,
+                                      start_at: start_time,
+                                      end_at: nil,
+                                      created_at: start_time,
+                                      updated_at: start_time,
+                                      total_time: 0
+      @store.save(new)
+    end
+  end
+
+  # Record the start time and clear the end time for
+  # every entry with the (map_id, round) combination.
+  def start_timing_for_round(map_id, round)
+    # check for pre-existing record
+    records = @store.where(
+      map_id: map_id,
+      round: round
+    )
+
+    _record_start_time(records)
+  end
+
+  # Actually does the work of recording the start
+  # time for the set of LocalSubmittedContent records
+  # provided.
+  def _record_start_time(records)
+    start_time = DateTime.now
+    unless records.empty?
+      records.each do |record|
+        record.start_at = start_time
+        record.end_at = nil
+        record.updated_at = start_time
+      end
+    end
+  end
+
+  # End timing for a single [link] in the given review and [round].
+  def end_timing_for_link(map_id, round, link)
+    # if link is provided, we'll update the end time for it
+    records = @store.where(map_id: map_id, round: round, link: link)
+    _record_end_time(records)
+  end
+
+  # End timing for all links for the given [map_id] and [round].
+  def end_timing_for_round(map_id, round)
+    # if no specific link is provided, then update the end
+    # time for all links
+    records = @store.where(map_id: map_id, round: round)
+    _record_end_time(records)
+  end
+
+  # For each record in [records], set the end time
+  # to DateTime.now and update the total_time accumulator
+  def _record_end_time(records)
+    end_time = DateTime.now
+    unless records.empty?
+      records.each do |record|
+        record.end_at = end_time
+        record.updated_at = end_time
+        record.total_time += record.time_diff
+      end
+    end
+  end
+
+  # Actually performs the work flushing records from
+  # local storage to the database and removing them.
+  def save_and_remove_all(map_id, round)
+    uncommitted = []
+    records = @store.where(map_id: map_id, round: round)
+
+    unless records.empty?
+      records.each do |record|
+        # push the uncommitted link on to the stack
+        uncommitted << record.link
+
+        previous = SubmissionViewingEvent.where(
+          map_id: record.map_id,
+          round: record.round,
+          link: record.link
+        )
+
+        if !previous.empty?
+          # make sure to add the total time on this record
+          # with what may have already been in the database
+          previous.each do |event|
+            updated = record.merge(event)
+            event.update_attribute(:total_time, updated)
+          end
```

Comment:
They might not like this due to nested loops. I solved this in my code bc often times the code inside the loop would be in it's own function with it's own loop

Reply:
Good feedback. I just refactored and cleaned up things on this one, didn't put a lot of thought in to restructuring the logic to avoid the nested loops.